Device encryption adiantum (#1342)
OWASP/owasp-mstg: Document/0x05a-Platform-Overview.md

    ...
```

#### Android Device Encryption

//INTRO

##### Full-Disk Encryption

//Full-Disk Encryption content

##### File-Based Encryption

//File-Based Encryption content

##### Adiantum 

AES is used on most modern Android devices for storage encryption. Actually, AES has become such a widely used algorithm that the most recent processor implementations have a dedicated set of instructions to provide hardware accelerated encryption and decryption operations, such as ARMv8 with its Cryptography Extensions or x86 with AES-NI extension.

> However, not all devices are capable of using AES for storage encryption in a timely fashion. Especially those that use Android Go (an Android version targeted at low-end devices), smartwatches or TV's. These devices usually use low-end processors, such as the ARM Cortex-A7 which don't have hardware accelerated AES.

Reviewer: Can you change `Especially those that use Android Go (an Android version targeted at low-end devices), ` to `Especially low-end devices running Android Go.`

> Android 5.0 and above support full-disk encryption. This encryption uses a single key protected by the users' device password to cipher the userdata partition. This kind of encryption is now considered  deprecated and file-based encryption should be used whenever possible. Full-disk encryption has drawbacks, such as not being able to receive calls or not having operative alarms after a reboot if the user does not enter its password. 

Reviewer: ```suggestion
Android 5.0 and above support full-disk encryption. This encryption uses a single key protected by the users' device password to encrypt & decrypt the userdata partition. This kind of encryption is now considered deprecated and file-based encryption should be used whenever possible. Full-disk encryption has drawbacks, such as not being able to receive calls or not having operative alarms after a reboot if the user does not enter its password. 
```

> Android 5.0 and above support full-disk encryption. This encryption uses a single key protected by the users' device password to encrypt & decrypt the userdata partition. This kind of encryption is now considered deprecated and file-based encryption should be used whenever possible. Full-disk encryption has drawbacks, such as not being able to receive calls or not having operative alarms after a reboot if the user does not enter its password. 

Reviewer 2: ```suggestion
Android 5.0 and above support full-disk encryption. This encryption uses a single key protected by the users' device password to encrypt and decrypt the userdata partition. This kind of encryption is now considered deprecated and file-based encryption should be used whenever possible. Full-disk encryption has drawbacks, such as not being able to receive calls or not having operative alarms after a reboot if the user does not enter his password. 
```

> Android 7.0 supports file-based encryption. File-based encryption allows different files to be encrypted with different keys so they can be deciphered independently. Devices that support this encryption also support Direct Boot which enables the device to have access to features such as alarms or accessibility services even if the user does not enter its password.

Reviewer: ```suggestion
Android 7.0 supports file-based encryption. File-based encryption allows different files to be encrypted with different keys so they can be deciphered independently. Devices which support this type of encryption support Direct Boot as well. Direct Boot enables the device to have access to features such as alarms or accessibility services even if the user does not enter its password.
```

> Android 7.0 supports file-based encryption. File-based encryption allows different files to be encrypted with different keys so they can be deciphered independently. Devices which support this type of encryption support Direct Boot as well. Direct Boot enables the device to have access to features such as alarms or accessibility services even if the user does not enter its password.

Reviewer 2: ```suggestion
Android 7.0 supports file-based encryption. File-based encryption allows different files to be encrypted with different keys so they can be deciphered independently. Devices which support this type of encryption support Direct Boot as well. Direct Boot enables the device to have access to features such as alarms or accessibility services even if the user does not enter his password.
```

Adiantum is a cipher construction designed by Paul Crowley and Eric Biggers at Google to fill the gap for that set of devices which are not able to run AES at least at 50 MiB/s. Adiantum relies only on additions, rotations and XORs; these operations are natively supported on all processors. Therefore, the low-end processors can encrypt 4 times faster and decrypt 5 times faster than they would if they were using AES.

Adiantum is a composition of other ciphers: 
- NH: A hashing function.
- Poly1305: A message authentication code (MAC).
- XChaCha12: A stream cipher.
- AES-256: A single invocation of AES.

Adiantum is a new cipher but it is secure, as long as ChaCha12 and AES-256 are considered secure. Its designers didn't create any new cryptographic primitive, instead they relied on other well-known and thoroughly studied primitives to create a new performant algorithm.

Adiantum is available for Android 9 and higher versions. However, modifications to the kernel and userspace have to be carried out.

> Android does not provide an API to application developers to use Adiantum; this cipher is to be taken into account and implemented by ROM developers or device vendors, which want to provide full disk encryption without sacrificing performance on low-end devices.

Reviewer: Is there a native library supporting it already? I could not find one... Because that would be easier than doing firmware hacks. If there is none, maybe we can say something like it was not available at the time of writing, but we are looking for one?

Author: It is natively supported in Linux kernel 5.0 and onwards, kernel 4.19, 4.14 & 4.9 need patching.
I don't know of any user-space library which implements Adiantum. As per Google documentation it seems that this cipher is intended to be used at filesystem level only. So, I think that maybe there won't be much interest to implement in a crypto library.
I will clarify Adiantum availability in the different kernels and also that no library implements it at the moment.

> Adiantum is available for Android 9 and higher versions. It is natively supported in Linux kernel 5.0 and onwards, kernel 4.19, 4.14 & 4.9 need patching. 

Reviewer: ```suggestion
Adiantum is available for Android 9 and higher versions. It is natively supported in Linux kernel 5.0 and onwards, while kernel 4.19, 4.14 & 4.9 need patching. 
```

> It should be noted that AES runs faster on those devices that have the AES instruction set. In that case the use of Adiantum is highly discouraged.

Reviewer 2: ```suggestion
It should be noted that AES runs faster on devices having the AES instruction set. In that case the use of Adiantum is highly discouraged.
```
